### ISSUE TITLE
Delete model: Volume memory leak

### ISSUE THREAD

**Issue by tbiedert** [ospray/ospray#112]
I just noticed volume memory is not freed when deleting both model and volume (see code below). You have to manually remove the volume from the model first.

The volume's refcount should be reduced automatically when deleting the model.

Another note: What about an automatic ospRelease() call in each destructor of the C++ interface?

```
/*
 * Create volume and model
 */
volume = new ospray::cpp::Volume("block_bricked_volume");
// ... set parameters
volume->commit();

model = new ospray::cpp::Model();
model->addVolume(*volume);
model->commit();


/*
 * Later: clean up again
 */
if (model != nullptr)
{
    model->removeVolume(*volume);     // <--- THIS IS NECESSARY, OTHERWISE VOLUME MEMORY WILL NOT BE FREED
    ospRelease(model->handle());
    delete model;
}

if (volume != nullptr)
{
    ospRelease(volume->handle());
    delete volume;
}
```

**Comment by jeffamstutz** [comment]
OSPRay's memory management semantics were designed to be opaque, which means that internal references may be maintained without the application (i.e. user API's perspective) containing any references to an object. Thus the semantic meaning of ospRelease() is that the application no longer wants a reference to the object's handle, not that the object will, in-fact, be freed/deleted. It's subtle in meaning, but entirely intended behavior.

In the case of a volume in a model, when the volume was added to the model, the model internally creates a reference to it. Thus even if the volume is released by the application, the model which contains a reference to it remains valid. Furthermore (at least I think this is true), you may have noticed that if you release the model too, it isn't freed either because you have to add the model to the renderer you want to use; the same principle applying here.

As for the C++ wrappers, there definitely is a desire to have ospRelease() get called correctly, but that requires refcounts in the C++ wrappers themselves. That is because having two C++ objects (e.g. two ospray::cpp::Volume objects), if you do an assignment from one to the other, it (roughly) behaves like a std::shared_ptr in that you don't get a deep copy of the volume: you only get a copy of the object handle. Because ospRelease() expresses that the *entire* application is done with the object, there needs to be code which tracks if there are any outstanding C++ objects the app has. Thus if you always do ospRelease() in the destructor of the C++ class, then it invalidates any C++ wrapper copies within the app, which isn't the semantics I intended with the wrappers. Totally solvable, it just needs to be prioritized.

In general, memory management is less than perfect right now in OSPRay, but we don't want that characteristic to last forever. I think I share your general position on wanting memory correctness to be perfect, but the lack of our user's feedback regarding the issue means it simply hasn't been prioritized above other things....thanks for the concrete "request" :)

I hope that helps!

Cheers,
Jeff

**Comment by timbiedert** [comment]
Thank you for the the detailed answer! 

Regarding the model-renderer relationship:   How can I remove the model from the renderer? ManagedObject_T<>::set() requires a reference as parameter.

In general, if I have constructed my full OSPRay "setup", where transfer function belongs (= is set) to the volume, volume to model, model and camera to renderer etc., what's the proper way to completely destroy everything again?

**Comment by jeffamstutz** [comment]
Hmm, right now there isn't a good way to remove a parameter. Right now you could exploit a hack where setting any parameter with the same string will clear the previous parameter, even it is a different type. Thus you could set an int parameter called "model" on the renderer, which would remove the reference to the model held by the renderer. What this really begs for is a generic ospRemoveParam() (w/ corresponding C++ method) function in the API.

The best way to destroy everything is to simply ospRelease() everything the API has given you. I *think* that if there are no more external references to any object, then their internal destructors should start being invoked, where the destruction of higher level objects (i.e. a renderer) will remove references to lower level objects, causing a chain-reaction of destructors. If this turns out to *not* be the case for you, that is a bug which needs fixed!

**Comment by jeffamstutz** [comment]
FYI, we are now tracking both features: 1) C++ ospRelease() correctness, and 2) some sort of ospRemoveParam() function.

**Comment by jeffamstutz** [comment]
FYI, ospRemoveParam() was implemented in c73b14be and 1cce5ba0 (both currently in ```devel```). It is also exposed in the C++ wrappers and scripting environment too.

**Comment by jeffamstutz** [comment]
UPDATE: "smart" handle management has been bumped to v1.3.0.

**Comment by paulmelis** [comment]
Sorry for hijacking this issue, but this paragraph caught my attention:

> OSPRay's memory management semantics were designed to be opaque, which means that internal references may be maintained without the application (i.e. user API's perspective) containing any references to an object. Thus the semantic meaning of ospRelease() is that the application no longer wants a reference to the object's handle, not that the object will, in-fact, be freed/deleted. It's subtle in meaning, but entirely intended behavior.

So that sounds like ospRelease() is semantically quite different from just decreasing a refcount 
 to an object (e.g. like Py_DECREF() in the Python C API, which I have used a lot). What does that mean for a piece of code like this (basically copied from ospTutorial.cpp)?

```OSPData data = ospNewData(num_vertices, OSP_FLOAT3, vertices);    
ospCommit(data);
ospSetData(mesh, "vertex", data);

data = ospNewData(num_vertices, OSP_FLOAT4, colors);
ospCommit(data);
ospSetData(mesh, "vertex.color", data);
```

Does it make sense to call ospRelease() on the `data` variable before creating the color array? Or shouldn't I worry?

**Comment by jeffamstutz** [comment]
@paulmelis sorry for not seeing your response such a long time ago. By not calling ospRelease() before overwriting with the second ```ospNewData()``` call, you leak the former data array because OSPRay will always think you want to keep that handle around to do something else with it (e.g. use it on another object), but you'll no longer have a handle to it to later call ```ospRelease()```.

I was trying to implement the proposed fix for ```v1.4.0```, but as it turns out this is difficult to get right. The reason is that when a C++ wrapper instance goes out of scope, we don't know if it is because of normal C++ scoping rules or if the application is being closed (process terminate). That is important to know when the currently set ```OSPDevice``` instance is freed (internally held by a refcounted pointer) before or after the eventual calls to ```ospRelease()``` (done internally in the C++ wrappers).

Thus I think this is actually a no-go on implementing, as we want to rely on users _not_ calling ```ospRelease()``` on app shutdown so we don't have to enforce destructor ordering... The solution will simply be that API users will have to track when the application is no longer interested in holding on to a handle.

I think when we get to OSPRay ```v2.0```, which is when we want to introduce any breaking changes to the API, we will very much rethink this so you don't have tediously track when ```ospRelease()``` should be called. Fortunately in the mean time we have been developing a scene graph (```ospray_sg```) that we want to keep as a stable API which can do this tracking for you. Documentation and features in that will be coming soon (sometime after ```v1.4.0```), but an easier API to OSPRay that adds more robust checking around things like parameters and handle lifetimes is coming...
